tests/test_convert.py

```diff
@@ -840,3 +926,28 @@ def test_alias_rejects_bad_types():
     """
     with pytest.raises(ConversionError, match="aliases must be of registers"):
         parse(source)
+
+
+def test_reject_mixed_addressing_mode():
+    source1 = """
+    qubit q1;
+    reset $0;
+    """
+    with pytest.raises(
+        ConversionError, match="Physical qubit referenced in virtual addressing mode"
+    ):
+        parse(source1)
+
+    source2 = """
+    reset $0;
+    qubit q1;
+    """
+    with pytest.raises(ConversionError, match="Virtual qubit declared in physical addressing mode"):
+        parse(source2)
+
+    source3 = """
+    reset $0;
+    qubit[3] q;
+    """
+    with pytest.raises(ConversionError, match="Virtual qubit declared in physical addressing mode"):
+        parse(source3)
```

Comment:
I'd marginally prefer to separate out the tests, so they all get tested individually rather than failing at the first.

Please can we also add some failure (and success) tests that we get suitable behaviour if the hardware qubits are only used within nested scopes, etc.

Hardware qubits should also be rejected from within `gate` bodies, regardless of the mode - gates can't reference them:

> The statements in the body can only refer to the symbols given in the parameter or argument list, and these symbols are scoped only to the subroutine body.

`gate` declarations should still work and be usable in hardware-qubit mode.